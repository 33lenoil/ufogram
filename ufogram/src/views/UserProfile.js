import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import {useLocation} from 'react-router-dom';
import getUser from '../api/getUser';

export default function Main() {
    const navigate = useNavigate();

    const location = useLocation();

    const [followed, setFollowed] = useState(false);

    const isMyself = location.state.self;

    const handleMain = () => {
        navigate('/main',  { state: { userId: location.state.userId, username: location.state.username, users: location.state.users } });
    };

    const handleFollow = (followEvent) => {
        if (followed) {
            followEvent.target.innerHTML = "Follow";
            setFollowed(false);
        } else {
            followEvent.target.innerHTML = "Unfollow";
            setFollowed(true);
        }
    };

    const GetFollowing = () => {
        const [following, setFollowing] = useState([]);

        useEffect(() => {
            async function fetchFollowing() {
                try {
                    console.log(location.state.sId);
                    const response = await getUser(location.state.sId);
                    if (response.status === 200) {
                        console.log(response.data.following);
                        setFollowing(response.data.following);
                    } else {
                        // Authentication failed, set error message
                        console.log('Invaliduser ID. Please try again.');
                    }
                } catch (error) {
                    console.error('getUser error', error);
                }
            }
            fetchFollowing();
        }, []);
        return following;
    }

    const GetFollowers = () => {
        const [followers, setFollowers] = useState([]);

        useEffect(() => {
            async function fetchFollowers() {
                try {
                    console.log(location.state.sId);
                    const response = await getUser(location.state.sId);
                    if (response.status === 200) {
                        console.log(response.data.followers);
                        setFollowers(response.data.followers);
                    } else {
                        // Authentication failed, set error message
                        console.log('Invaliduser ID. Please try again.');
                    }
                } catch (error) {
                    console.error('getUser error', error);
                }
            }
            fetchFollowers();
        }, []);
        return followers;
    }

    return (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
            <div style={{display: "flex", justifyContent: "center", position: "fixed", width: "100%", background: "#8769b6"}}>
                <h1>Profile</h1>
                <div>
                    <button type="button" title="Create New Post" onClick={handleMain}>Main</ button>
                </div>
            </div>
            <div style={{display: "flex", width: "100%", justifyContent: "center", marginTop: "100px", background: "#b6f486"}}>
                <div style={{display: "flex", width: "100%", justifyContent: "space-between", flexDirection: "row"}}>
                    <div style={{display: "flex", justifyContent: "space-between", flexDirection: "row"}}>
                        <label for="followers">Followers</label>
                        <select name="followers" id="followers">
                        {GetFollowers().map(user => (
                        <option value={ user.username }>{ user.username }</option>
                        ))
                        }
                        </select>
                    </div>
                    <div style={{display: "flex", flexDirection: "column"}}>
                        <h2>{ location.state.sName }</h2>
                        <img src="https://picsum.photos/200/304" />
                        <t>My info</t>
                        { isMyself !== true &&
                            <button type="button" title="Follow/Unfollow" onClick={handleFollow}>Follow</button>
                        }
                    </div>
                    <div style={{display: "flex", justifyContent: "space-between", flexDirection: "row"}}>
                        <label for="following">Following</label>
                        <select name="following" id="following">
                        {GetFollowing().map(user => (
                        <option value={ user.username }>{ user.username }</option>
                        ))
                        }
                        </select>
                    </div>
                </div>  
            </div>
        </div>
    )
}