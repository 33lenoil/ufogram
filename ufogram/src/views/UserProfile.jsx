import React, { useEffect, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import getUser from '../api/getUser';
import updateUser from '../api/updateUser';

export default function UserProfile() {
  const navigate = useNavigate();

  const location = useLocation();

  // const [followed, setFollowed] = useState(location.state.followed);
  let { followed } = location.state;

  const isMyself = location.state.self;

  const handleMain = () => {
    navigate('/main', { state: { userId: location.state.userId, username: location.state.username, users: location.state.users } });
  };

  const handleFollow = async (followEvent) => {
    // console.log(`Followed at the start: ${followed}`);
    const selfResponse = await getUser(location.state.userId);
    const searchResponse = await getUser(location.state.sId);
    const currFollowing = selfResponse.data.following.slice();
    const currFollowers = searchResponse.data.followers.slice();
    const copy = followEvent;
    if (followed) {
      const newFollowing = currFollowing.filter((user) => user.id !== location.state.sId);
      const newFollowers = currFollowers.filter((user) => user.id !== location.state.userId);
      // console.log('newFollowing: \n', newFollowing);
      // console.log('newFollowers: \n', newFollowers);
      await updateUser(location.state.userId, {
        username: selfResponse.data.username,
        password: selfResponse.data.password,
        following: newFollowing,
        followers: selfResponse.data.followers,
      });
      await updateUser(location.state.sId, {
        username: searchResponse.data.username,
        password: searchResponse.data.password,
        following: searchResponse.data.following,
        followers: newFollowers,
      });
      copy.target.innerHTML = 'Follow';
      // setFollowed(false);
      followed = false;
      // console.log(`Shoudl be false: ${followed}`);
    } else {
      const newFollowing = currFollowing.slice();
      const newFollowers = currFollowers.slice();
      newFollowing.push({
        id: location.state.sId,
        username: searchResponse.data.username,
      });
      newFollowers.push({
        id: location.state.userId,
        username: selfResponse.data.username,
      });
      // console.log('newFollowing: \n', newFollowing);
      // console.log('newFollowers: \n', newFollowers);
      await updateUser(location.state.userId, {
        username: selfResponse.data.username,
        password: selfResponse.data.password,
        following: newFollowing,
        followers: selfResponse.data.followers,
      });
      await updateUser(location.state.sId, {
        username: searchResponse.data.username,
        password: searchResponse.data.password,
        following: searchResponse.data.following,
        followers: newFollowers,
      });

      copy.target.innerHTML = 'Unfollow';
      // console.log('Inner HTML: ', followEvent.target.innerHTML);
      // setFollowed(true);
      followed = true;
      // console.log(`Shoudl be true: ${followed}`);
    }
  };

  const GetFollowing = () => {
    const [following, setFollowing] = useState([]);

    useEffect(() => {
      async function fetchFollowing() {
        try {
          // console.log(location.state.sId);
          const response = await getUser(location.state.sId);
          if (response.status === 200) {
            // console.log(response.data.following);
            setFollowing(response.data.following);
          } else {
            // Authentication failed, set error message
            // console.log('Invaliduser ID. Please try again.');
          }
        } catch (error) {
          // console.error('getUser error', error);
        }
      }
      fetchFollowing();
    }, []);
    return following;
  };

  const GetFollowers = () => {
    const [followers, setFollowers] = useState([]);

    useEffect(() => {
      async function fetchFollowers() {
        try {
          // console.log(location.state.sId);
          const response = await getUser(location.state.sId);
          if (response.status === 200) {
            // console.log(response.data.followers);
            setFollowers(response.data.followers);
          } else {
            // Authentication failed, set error message
            // console.log('Invaliduser ID. Please try again.');
          }
        } catch (error) {
          // console.error('getUser error', error);
        }
      }
      fetchFollowers();
    }, []);
    return followers;
  };

  return (
    <div style={{ display: 'flex', justifyContent: 'space-evenly' }}>
      <div style={{
        display: 'flex', justifyContent: 'center', position: 'fixed', width: '100%', background: '#8769b6',
      }}
      >
        <h1>Profile</h1>
        <div>
          <button type="button" title="Create New Post" onClick={handleMain}>Main</button>
        </div>
      </div>
      <div style={{
        display: 'flex', width: '100%', justifyContent: 'center', marginTop: '100px', background: '#b6f486',
      }}
      >
        <div style={{
          display: 'flex', width: '100%', justifyContent: 'space-between', flexDirection: 'row',
        }}
        >
          <div style={{ display: 'flex', justifyContent: 'space-between', flexDirection: 'row' }}>
            <label htmlFor="followers">
              Followers
              <select name="followers" id="followers">
                {GetFollowers().map((user) => (
                  <option value={user.username}>{ user.username }</option>
                ))}
              </select>
            </label>
          </div>
          <div style={{ display: 'flex', flexDirection: 'column' }}>
            <h2>{ location.state.sName }</h2>
            <img alt="test profile pic" src="https://picsum.photos/200/304" />
            <t>My info</t>
            { isMyself !== true && followed === true
                            && <button type="button" title="Follow/Unfollow" onClick={handleFollow}>Unfollow</button>}
            { isMyself !== true && followed === false
                            && <button type="button" title="Follow/Unfollow" onClick={handleFollow}>Follow</button>}
          </div>
          <div style={{ display: 'flex', justifyContent: 'space-between', flexDirection: 'row' }}>
            <label htmlFor="following">
              Following
              <select name="following" id="following">
                {GetFollowing().map((user) => (
                  <option value={user.username}>{ user.username }</option>
                ))}
              </select>
            </label>
          </div>
        </div>
      </div>
    </div>
  );
}
