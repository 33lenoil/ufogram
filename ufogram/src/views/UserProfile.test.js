/**
 * @jest-environment jsdom
 */

import React, { useState } from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import UserProfile from './UserProfile';

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => ({
      pathname: "localhost:3001/userprofile",
      state: { userId: 2, username: 'lionelhu', self: false, sName: 'zairuiy', sId: 1, users: [], followed: false }
    })
  }));


test('renders title', () => {
    render(
      <Router>
          {/* <Routes>
              <Route path="/signup" element={<Login />} />
          </Routes> */}
          <UserProfile />
      </Router>
    );
    const linkElement = screen.getByText(/Profile/);
    expect(linkElement).toBeInTheDocument();
  });

test('renders main button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <UserProfile />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Main/
  })
  expect(linkElement).toBeInTheDocument();
});

test('renders follow button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <UserProfile />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Follow/
  })
  expect(linkElement).toBeInTheDocument();
});

test('renders my info text', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <UserProfile />
        </Router>
      );
      const linkElement = screen.getByText(/My info/);
      expect(linkElement).toBeInTheDocument();
  });

  test('renders name', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <UserProfile />
        </Router>
      );
      const linkElement = screen.getByText(/zairuiy/);
      expect(linkElement).toBeInTheDocument();
  });

test('renders followers label', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Signup />} />
            </Routes> */}
            <UserProfile />
        </Router>
      );
      const linkElement = screen.getByText(/Followers/);
    expect(linkElement).toBeInTheDocument();
});

test('renders following label', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Signup />} />
            </Routes> */}
            <UserProfile />
        </Router>
      );
      const linkElement = screen.getByText(/Following/);
    expect(linkElement).toBeInTheDocument();
});

/**
 * Snapshot Testing
 */

test('the component matches the snapshot', () => {
  const component = renderer.create(<Router>
    {/* <Routes>
        <Route path="/signup" element={<Signup />} />
    </Routes> */}
    <UserProfile />
</Router>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// test('welcome is displayed after click the registration button', async () => {
//   render(<Login />);
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByLabelText(/welcome/i);
//   expect(welcomeElement).toBeInTheDocument();
// });

// test('Username is displayed after clicking on registration button', async () => {
//   render(<Login />);
//   const usernameElement = screen.getByTestId('uname');
//   await userEvent.type(usernameElement, 'lili');
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByText(/lili/i);
//   expect(welcomeElement).toBeInTheDocument();
// });
