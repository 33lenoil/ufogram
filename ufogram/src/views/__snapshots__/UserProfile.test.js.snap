// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component matches the snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-evenly",
    }
  }
>
  <div
    style={
      Object {
        "background": "#8769b6",
        "display": "flex",
        "justifyContent": "center",
        "position": "fixed",
        "width": "100%",
      }
    }
  >
    <h1>
      Profile
    </h1>
    <div>
      <button
        onClick={[Function]}
        title="Create New Post"
        type="button"
      >
        Main
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "background": "#b6f486",
        "display": "flex",
        "justifyContent": "center",
        "marginTop": "100px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <label
          htmlFor="followers"
        >
          Followers
        </label>
        <select
          id="followers"
          name="followers"
        />
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      >
        <h2>
          zairuiy
        </h2>
        <img
          src="https://picsum.photos/200/304"
        />
        <t>
          My info
        </t>
        <button
          onClick={[Function]}
          title="Follow/Unfollow"
          type="button"
        >
          Follow
        </button>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <label
          htmlFor="following"
        >
          Following
        </label>
        <select
          id="following"
          name="following"
        />
      </div>
    </div>
  </div>
</div>
`;
